internal/controllers/upboundagent/reconciler.go

```diff
@@ -0,0 +1,182 @@
+)
+
+const (
+	errGetSecret              = "failed to get control plane token secret"
+	errNoTokenInSecret        = "secret %s does not contain a token for key \"%s\""
+	errFailedToSyncDeployment = "failed to sync agent deployment"
+)
+
+var (
+	secretsKind     = schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Secret"}
+	deploymentsKind = schema.GroupVersionKind{Group: "apps", Version: "v1", Kind: "Deployment"}
+)
+
+// ReconcilerOption is used to configure the Reconciler.
+type ReconcilerOption func(*Reconciler)
+
+// WithLogger specifies how the Reconciler should log messages.
+func WithLogger(log logging.Logger) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.log = log
+	}
+}
+
+// Setup adds a controller that reconciles on control plane token secret and manages Upbound Agent deployment
+func Setup(mgr ctrl.Manager, l logging.Logger, ds appsv1.DeploymentSpec, ts string) error {
+	name := "upboundAgent"
+
+	r := NewReconciler(mgr, ds, ts,
+		WithLogger(l.WithValues("controller", name)),
+	)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&corev1.Secret{}).
+		Owns(&appsv1.Deployment{}).
+		WithEventFilter(resource.NewPredicates(resource.AnyOf(
+			resource.AllOf(IsOfKind(secretsKind, mgr.GetScheme()), resource.IsNamed(ts)),
+			resource.AllOf(IsOfKind(deploymentsKind, mgr.GetScheme()), resource.IsNamed(deploymentUpboundAgent)),
+		))).
+		Complete(r)
+}
+
+// Reconciler reconciles on control plane token secret and manages Upbound Agent deployment
+type Reconciler struct {
+	client         client.Client
+	deploymentSpec appsv1.DeploymentSpec
+	tokenSecret    string
+	log            logging.Logger
+}
+
+// NewReconciler returns a new reconciler
+func NewReconciler(mgr manager.Manager, ds appsv1.DeploymentSpec, ts string, opts ...ReconcilerOption) *Reconciler {
+	r := &Reconciler{
+		client:         mgr.GetClient(),
+		deploymentSpec: ds,
+		tokenSecret:    ts,
+		log:            logging.NewNopLogger(),
+	}
+
+	for _, f := range opts {
+		f(r)
+	}
+
+	return r
+}
+
+// Reconcile reconciles on control plane token secret and manages Upbound Agent deployment
+func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("request", req)
+
+	log.Debug("Reconciling...")
+	ctx, cancel := context.WithTimeout(ctx, reconcileTimeout)
+	defer cancel()
+
+	ts := &corev1.Secret{}
+	err := r.client.Get(ctx, types.NamespacedName{Name: req.Name, Namespace: req.Namespace}, ts)
+
+	// We are using owner references to get agent deployment deleted when token secret is deleted.
+	// Nothing to do here if token secret deleted.
+	if kerrors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	}
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, errGetSecret)
+	}
+
+	// Ensure secret has token
+	t := ts.Data[keyToken]
+	if string(t) == "" {
+		err := errors.Errorf(errNoTokenInSecret, r.tokenSecret, keyToken)
+		log.Info(err.Error())
```

Comment:
I would normally suggest using `Debug` logs here, but since we don't really have a great way of publishing events for this controller then maybe `Info` makes sense?

Reply:
Yeah, that's the motivation. 
This is a bit different than other crossplane reconcilers where we reconcile on custom types with n instances. Here, we are reconciling on `Secret` type which does not show events with kubectl commands and we are working on a single instance. It is more intuitive for a user to check bootstrapper logs if sth goes wrong rather than looking around for events.